turbo_alignment/trainers/dpo.py

```diff
@@ -271,6 +286,46 @@ def compute_loss(
         )
 
 
+@DPOLossRegistry.register(DPOLossesType.RDPO)
+class RDPOLoss(DPOLossRegistry):
+    def __init__(self, *args, beta: float = 0.1, alpha: float = 0, **kwargs) -> None:
+        self.beta = beta
+        self.alpha = alpha
+        super().__init__(*args, **kwargs)
+
+    def compute_loss(
+        self,
+        policy_chosen_logps: torch.FloatTensor,
+        policy_rejected_logps: torch.FloatTensor,
+        reference_chosen_logps: torch.FloatTensor,
+        reference_rejected_logps: torch.FloatTensor,
+        policy_best_decode_logps: torch.FloatTensor | None,
+        precomputed_margins: torch.FloatTensor | None,
+        chosen_lens: int| None,
+        rejected_lens: int| None,
+    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        pi_logratios = policy_chosen_logps - policy_rejected_logps
+        ref_logratios = reference_chosen_logps - reference_rejected_logps
+
+        logits = pi_logratios - ref_logratios
+
+        chosen_rewards = self.beta * (policy_chosen_logps - reference_chosen_logps).detach()
+        rejected_rewards = self.beta * (policy_rejected_logps - reference_rejected_logps).detach()
+
+        unscaled = self.beta * logits
+
+        if chosen_lens is not None and rejected_lens is not None:
+            unscaled -= self.alpha * rejected_lens - self.alpha * chosen_lens
```

Comment:
мне визуально так приятней и проще, но и так как у тебя ок
```suggestion
            unscaled += self.alpha * (chosen_lens  - rejected_lens)
```